src/main/java/bio/terra/service/dataset/SchemaValidationContext.java

```diff
@@ -0,0 +1,304 @@
+package bio.terra.service.dataset;
+
+import bio.terra.common.PdaoConstant;
+import bio.terra.common.ValidationUtils;
+import bio.terra.model.ColumnModel;
+import bio.terra.model.DatePartitionOptionsModel;
+import bio.terra.model.IntPartitionOptionsModel;
+import bio.terra.model.TableDataType;
+import bio.terra.model.TableModel;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.springframework.validation.Errors;
+
+/**
+ * SchemaValidationContext represents shared functionality between DatasetRequestValidator (used for
+ * creating dataset schemas) and DatasetSchemaUpdateValidator (used for updating dataset schemas).
+ */
+public class SchemaValidationContext {
+
+  private static final String PRIMARY_KEY = "PrimaryKey";
+
+  private final HashMap<String, HashSet<String>> tableColumnMap;
+  private final HashMap<String, HashSet<String>> tableArrayColumns;
+  private final HashSet<String> relationshipNameSet;
+  private final String fieldName;
+
+  SchemaValidationContext(String field) {
+    tableColumnMap = new HashMap<>();
+    tableArrayColumns = new HashMap<>();
+    relationshipNameSet = new HashSet<>();
+    fieldName = field;
+  }
```

Comment:
- instead of `HashMap` and `HashSet` for declared types, it's better to use `Map` and `Set` since those are the interface types (vs the implementation types)
- there's no difference between initializing a field in a constructor vs at its declaration, so I usually prefer field initialization when possible
- I would change the constructor parameter name to match the field name
```suggestion
  private final Map<String, Set<String>> tableColumnMap = new HashMap<>();
  private final Map<String, Set<String>> tableArrayColumns = new HashMap<>();
  private final Set<String> relationshipNameSet = new HashSet<>();
  private final String fieldName;

  SchemaValidationContext(String fieldName) {
    this.fieldName = fieldName;
  }
```